File: 1-250/3. Longest Substring Without Repeating Characters/solution.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length()  ;
        int rs = 0 ;
        int first = 0 ; 
        HashMap< Character , Integer> map = new HashMap<>() ; 
        for(int i = 0 ; i<n ; ++i){
            char c = s.charAt(i) ; 
            if( map.containsKey( c ) ){
                first = Math.max(  map.get(c) +1 , first  ) ; 
            }
            map.put( c, i ) ; 
            rs = Math.max( rs , i - first +1 ) ; 
        }
        
        return rs ; 
    }
}

File: 1-250/3. Longest Substring Without Repeating Characters/solution.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.size();
        vector<int> pos(128, -1);
        int mxLen = 0, previ = 0;
        for (int i = 0; i < n; i++) {
            if (pos[s[i]] != -1 && pos[s[i]] >= previ) {
                previ = pos[s[i]] + 1;
            } else {
                mxLen = max(mxLen, i - previ + 1);
            }
            pos[s[i]] = i;
        }
        return mxLen;
    }
};